### PR TITLE
Do not panic when sending events to closed emitter

### PR DESCRIPTION
**Pull request by t0yv0**
# Description

Fixes [#10431](https://github.com/pulumi/pulumi/issues/10431)

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/engine/events.go [hunk 1]
```diff
@@ -345,42 +356,68 @@ func makeStepEventStateMetadata(state *resource.State, debug bool) *StepEventSta
...
+			e, event)
+	}
+}
+
+func (e *eventEmitter) sendEvent(event Event) {
+	if e.isClosed() {
+		e.logUnableToSendEvent(event)
+		return
+	}
+	e.ch <- event
```
Comment (t0yv0) [inline pkg/engine/events.go:386]: Tiny possibility of a race here, might need to use a mutex after all.

**Comment by t0yv0** [comment]
This is not quite right (though it passes the repro), need to work out the synchronization logic better.

File: pkg/engine/events.go [hunk 2]
```diff
@@ -278,7 +285,11 @@ func queueEvents(events chan<- Event, buffer chan Event, done chan bool) {
 				if !ok {
 					// If the event source has been closed, flush the queue.
 					for _, e := range queue {
-						events <- e
+						if !emitter.isClosed() {
+							events <- e
```
Comment (t0yv0) [inline pkg/engine/events.go:289]: This isn't right but tricky. emitter.isClosed() is always true here. However if we don't guard here we have another panic in the repro. The events channel gets closed outside of the event emitter code.

File: pkg/engine/events.go [hunk 3]
```diff
@@ -225,31 +226,37 @@ func makeEventEmitter(events chan<- Event, update UpdateInfo) (eventEmitter, err
...
-	go queueEvents(events, buffer, done)
-
-	return eventEmitter{
-		done: done,
-		ch:   buffer,
-	}, nil
+	closed := new(int32)
+	*closed = 0
+	emitter := eventEmitter{
+		closed: closed,
```
Comment (AaronFriel) [inline pkg/engine/events.go:245]: Is there a reason that would be preferred over this?

```suggestion
	closed := 0
	emitter := eventEmitter{
		closed: &closed,
```

**Review by AaronFriel — COMMENTED** [review]
Having refactored the event emitter, would it make sense to:

1. Move it to a separate package to keep the fields private
2. Use a `defer func() { if r := recover(); r != nil { ... }}` function in `sendEvents` rather than guards that check if it's closed?

As you note, a mutex is needed or you'll race the "is closed" check race.

**Comment by t0yv0** [comment]
> Move it to a separate package to keep the fields private

That's not really possible in go due to import cycles, like this:

```
package github.com/pulumi/pulumi/pkg/v3/backend
        imports github.com/pulumi/pulumi/pkg/v3/backend/display
        imports github.com/pulumi/pulumi/pkg/v3/engine
        imports github.com/pulumi/pulumi/pkg/v3/engine/internal/event_emitter
        imports github.com/pulumi/pulumi/pkg/v3/engine: import cycle not allowed
package github.com/pulumi/pulumi/pkg/v3/backend
        imports github.com/pulumi/pulumi/pkg/v3/backend/display
        imports github.com/pulumi/pulumi/pkg/v3/engine
        imports github.com/pulumi/pulumi/pkg/v3/engine/internal/event_emitter
        imports github.com/pulumi/pulumi/pkg/v3/engine: import cycle not allowed
```

We'd need to introduce abstractions to break the cycles; that'd take the PR off-track for a very small marginal gain - I feel like won't tackle it here.

**Comment by t0yv0** [comment]
Recovering from a chan send panic sounds like a good idea in our situation.. I'll try that. It's a code smell for sure I'll have to put heavy comments on, but I don't think I can confidently get the lifetimes right because *something* is closing engine events chan on termination but stepExecutor keeps sending new diag events.

**Comment by Frassle** [comment]
I think a code smell here is fine for now. A lot of this logic is probably gonna get gutted and re-done as part of splitting the engine and cli up.

**Comment by t0yv0** [comment]
@Frassle @AaronFriel PTAL - the last version still passes my repro test.

**Review by AaronFriel — APPROVED** [review]
lgtm, I appreciate scoping & documenting the defer/recover block in `trySendEvent`

**Pull request merged**